AlphaWallet/KeyManagement/EtherKeystore.swift

```diff
@@ -0,0 +1,647 @@
+                .flatMap { decryptPrivateKey(fromCipherTextData: $0, forAccount: account, withUserPresence: withUserPresence) }
+        
+        //We copy the record that doesn't require user-presence make a new one which requires user-presence and read from that. We don't want to read the one without user-presence unless absolutely necessary (e.g user has disabled passcode)
+        if data == nil && withUserPresence && shouldWriteWithUserPresenceIfNotFound && keychain.lastResultCode == errSecItemNotFound {
+            if let keyWithoutUserPresence = getPrivateKeyFromNonHdWallet(forAccount: account, prompt: prompt, withUserPresence: false, shouldWriteWithUserPresenceIfNotFound: false) {
+                savePrivateKeyForNonHdWallet(keyWithoutUserPresence, forAccount: account, withUserPresence: true)
+            }
+            return getPrivateKeyFromNonHdWallet(forAccount: account, prompt: prompt, withUserPresence: true, shouldWriteWithUserPresenceIfNotFound: false)
+        } else {
+            return data
+        }
+    }
+
+    private func getSeedPhraseForHdWallet(forAccount account: EthereumAccount, prompt: String, withUserPresence: Bool) -> String? {
+        return getSeedForHdWallet(forAccount: account, prompt: prompt, withUserPresence: withUserPresence)
+                .flatMap { HDWallet(seed: $0, passphrase: emptyPassphrase) }
+                .flatMap { $0.mnemonic }
+    }
+
+    private func getSeedForHdWallet(forAccount account: EthereumAccount, prompt: String, withUserPresence: Bool, shouldWriteWithUserPresenceIfNotFound: Bool = true) -> String? {
+        let prefix: String
+        if withUserPresence {
+            prefix = Keys.ethereumSeedUserPresenceRequiredPrefix
+        } else {
+            prefix = Keys.ethereumSeedUserPresenceNotRequiredPrefix
+        }
+        let data = keychain.getData("\(prefix)\(account.address.eip55String)", prompt: prompt)
+                .flatMap { decryptHdWalletSeed(fromCipherTextData: $0, forAccount: account, withUserPresence: withUserPresence) }
+                .flatMap { String(data: $0, encoding: .utf8) }
+        //We copy the record that doesn't require user-presence make a new one which requires user-presence and read from that. We don't want to read the one without user-presence unless absolutely necessary (e.g user has disabled passcode)
+        if data == nil && withUserPresence && shouldWriteWithUserPresenceIfNotFound && keychain.lastResultCode == errSecItemNotFound {
+            if let seedWithoutUserPresence = getSeedForHdWallet(forAccount: account, prompt: prompt, withUserPresence: false, shouldWriteWithUserPresenceIfNotFound: false) {
+                saveSeedForHdWallet(seedWithoutUserPresence, forAccount: account, withUserPresence: true)
+            }
+            return getSeedForHdWallet(forAccount: account, prompt: prompt, withUserPresence: true, shouldWriteWithUserPresenceIfNotFound: false)
```

Comment:
A little confused by the flow here. Why are we reading, writing, and then re-reading and returning what we just wrote? Is that to ensure the writing was successful? A more natural flow would be to simply return the seed we read the first time around.

Reply:
I'll reuse the notation k to refer to the ECDSA key that requires user presence to access and k' to refer to the one that doesn't require user presence and just omit the corresponding ciphertext (they are accessed at the same time as k and k'):

We always try to access the wallet via k. This attempt to access the data will cause iOS the prompt the user to verify user presence.

In the line preceding the if-else, k is accessed. The `if` checks that access has failed because the k doesn't exist (instead of the user cancelling the operation or failing to prove user presence), so we attempt to read k' and write k then try to read k again.

The `else` only gets accessed if user-presence is not required.

Reply:
Thanks for explaining. This explanation might also be valuable to add as a code comment.